### ISSUE TITLE
[bug] Effects of `config_options` leak between host/build contexts when a package is present in both

### ISSUE THREAD

**Issue by AlexRamallo** [conan-io/conan#11385]
### Environment Details (include every applicable attribute)
  * Operating System+version: Ubuntu
  * Conan version: 1.49.0
  * Python version: 3.9.7

### Steps to reproduce (Include if Applicable)

The libpng package on Conan Center can illustrate this behavior. There's a reproduction [available here](https://github.com/AlexRamallo/conan_bug).

It seems as though adding the same package to both the build and host contexts causes some of the effects of `config_options` to affect both during install. In the repro above, libpng is in both the `requires` and `tool_requires` sections of the conanfile.

In [config_options](https://github.com/conan-io/conan-center-index/blob/5791577e80fa4724abc8deb04fa0cb4376d02ba9/recipes/libpng/all/conanfile.py#L67), the libpng recipe selectively deletes options based on capabilities (e.g. the `neon` option is deleted if arch isn't ARM). This causes the error shown in the logs below.

I assume it's because the option got deleted in the build context (x86_64) first, and when the host context began getting processed, some part of the recipe tried to access that deleted option. Idk where exactly, maybe during processing of `default_options`? The error message isn't helpful. I attached a backtrace from pdb in Part 2 of the logs below after breaking in the error handler.

The issue can be fixed by changing the way options are removed in the libpng recipe (e.g. set neon to False instead of deleting it), but I think this is a bug with Conan since what that recipe is doing is what it the official docs recommend.

Also, someone recently mentioned this exact same issue with libpng on an [old issue from 2017](https://github.com/conan-io/conan/issues/1411#issuecomment-1048623427)

### Logs (Executed commands with output) (Include/Attach if Applicable)

# Part 1:
```
$ conan install . -if=build --build=libpng --profile:host=android --profile:build=default

Configuration (profile_host):
[settings]
arch=armv7hf
compiler=clang
compiler.libcxx=libc++
compiler.version=14
os=Android
os.api_level=21
[options]
[build_requires]
*: android-ndk/r24
[env]

Configuration (profile_build):
[settings]
arch=x86_64
build_type=Debug
compiler=gcc
compiler.cppstd=17
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux
[options]
[build_requires]
[env]

libpng/1.6.37: Forced build from source
ERROR: libpng/1.6.37: option 'neon' doesn't exist
Possible options are ['shared', 'fPIC', 'sse', 'api_prefix']
```

# Part 2
backtrace from pdb
```
  /home/aramallo/py3/bin/conan(8)<module>()
-> sys.exit(run())
  /home/aramallo/py3/lib/python3.9/site-packages/conans/conan.py(11)run()
-> main(sys.argv[1:])
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/command.py(2399)main()
-> error = command.run(args)
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/command.py(2216)run()
-> method(args[0][1:])
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/command.py(523)install()
-> info = self._conan.install(path=args.path_or_reference,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/conan_api.py(93)wrapper()
-> return f(api, *args, **kwargs)
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/conan_api.py(609)install()
-> deps_install(app=self.app,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/manager.py(58)deps_install()
-> deps_graph = graph_manager.load_graph(ref_or_path, create_reference, graph_info, build_modes,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_manager.py(127)load_graph()
-> deps_graph = self._resolve_graph(root_node, profile_host, profile_build, graph_lock,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_manager.py(289)_resolve_graph()
-> deps_graph = self._load_graph(root_node, check_updates, update,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_manager.py(410)_load_graph()
-> self._recurse_build_requires(graph, builder, check_updates, update, build_mode,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_manager.py(386)_recurse_build_requires()
-> _recurse_build_requires(br_build, profile_build.build_requires)
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_manager.py(369)_recurse_build_requires()
-> nodessub = builder.extend_build_requires(graph, node, br_list, check_updates,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_builder.py(104)extend_build_requires()
-> self._expand_require(br, node, graph, check_updates, update,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_builder.py(275)_expand_require()
-> self._expand_node(new_node, graph, new_reqs, node.ref, new_options, check_updates,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_builder.py(124)_expand_node()
-> new_options, new_reqs = self._get_node_requirements(node, graph, down_ref, down_options,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_builder.py(195)_get_node_requirements()
-> new_options = self._config_node(node, down_ref, down_options)
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/graph/graph_builder.py(369)_config_node()
-> run_configure_method(conanfile, down_options, down_ref, ref)
  /home/aramallo/py3/lib/python3.9/site-packages/conans/client/conanfile/configure.py(20)run_configure_method()
-> conanfile.options.propagate_upstream(down_options, down_ref, ref)
  /home/aramallo/py3/lib/python3.9/site-packages/conans/model/options.py(624)propagate_upstream()
-> self._package_options.propagate_upstream(option_values, down_ref, own_ref,
  /home/aramallo/py3/lib/python3.9/site-packages/conans/model/options.py(529)propagate_upstream()
-> self._ensure_exists(name)
  /home/aramallo/py3/lib/python3.9/site-packages/conans/model/options.py(445)_ensure_exists()
-> raise ConanException(option_not_exist_msg(field, list(self._data.keys())))
> /home/aramallo/py3/lib/python3.9/site-packages/conans/model/options.py(30)option_not_exist_msg()
-> return "\n".join(result)
```

**Comment by SpaceIm** [comment]
@memsharded @jcar87 I can reproduce this bug in conan-center in a recipe having qt in requirements & build requirements (libpng is a dependency of qt): https://github.com/conan-io/conan-center-index/pull/14723#issuecomment-1349838474

**Comment by memsharded** [comment]
I have been investigating this. 
The good news is that it seems to be working fine in Conan 2.0.

The bad news is that it doesn't seem to have a reasonable fix in 1.X, I have tried for a while, but couldn't find anything that was not breaking or with a very high risk of breaking.

The possible alternative would be to leverage the fact that options with ``None`` value are basically equivalent for most cases to not being defined. So instead of deleting the options, something like this in the ``configure()`` method will work:

```python
def test_both_context_options_error():
    # https://github.com/conan-io/conan/issues/11385
    c = TestClient()
    pkg = textwrap.dedent("""
        from conan import ConanFile
        class Pkg(ConanFile):
            name = "pkg"
            version = "0.1"
            settings = "arch"
            options = {"neon": [None, True, "check", False]}
            default_options = {"neon": True}

            def configure(self):
                if "arm" not in self.settings.arch:
                    self.options.neon = None
            """)
    c.save({"pkg/conanfile.py": pkg,
            "consumer/conanfile.py": GenConanfile().with_requires("pkg/0.1")
                                                   .with_build_requires("pkg/0.1")})
    c.run("create pkg ")
    c.run("create pkg -s arch=armv8")
    c.run("install consumer -s:b arch=x86_64 -s:h arch=armv8")
```

**Comment by memsharded** [comment]
I am proposing to close this issue as solved in 2.0, proved by this test: https://github.com/conan-io/conan/pull/12709
